Resolved SMS/internet submission status issue

jd-alexander opened the pull request:
Closes #2372

#### What has been done to verify that this works as intended?
Submissions were done via SMS and then using the internet. When the submission via the internet was completed the status was updated.

#### Why is this the best possible solution? Were any other approaches considered?
This is the best solution because at the moment the `SmsSubmissionManager` is used to store the progress and state of submissions being sent over SMS so if that state isn't supposed to be displayed then submissions have to be flagged so when a user is switching between transport types the upload list reacts accordingly.

#### Are there any risks to merging this code? If so, what are they?
No risks are associated. The behavior just needs to be verified.

#### Do we need any specific form for testing your changes? If so, please attach one.
No. Just use the `SMS test` form that's provided via the default Aggregate server.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploader.java:
```diff
@@ -153,6 +167,17 @@ public void setUploaderListener(InstanceUploaderListener sl) {
     }
 
+    private void deferSubmissionStatus(Iterator<String> instanceIds) {
+
```
Please remove this blank line. Is this being added by Android Studio? Can you turn it off?

Is there a particular reason to pass in the iterator? I think it would be clearer to call the set `instanceIds` and pass that in. Then the semantics are clear -- you're deferring submission status on the instanceIDs. Now the parameter name suggests that's what you want to do but actually you're passing in an iterator on the collection. Using a for each loop also makes things easier to read (that's why it's provided as syntactic sugar on top of iterators!).

```java
private void deferSubmissionStatus(Set<String> instanceIds) {
  for (String instanceId : instanceIds) {
    SmsSubmission model = submissionManager.getSubmissionModel(instanceId);
    ...
  }
}
```

jd-alexander replied:
Sure! Will do! I will check if Android Studio is actually adding it. Will also check my other PR as well as I think some of them are in there.

lognaturel commented:
If I understand correctly, you're always marking submissions as "deferred" after any send attempt and then unmarking them as "deferred" after an SMS send attempt. The deferral doesn't feel like it's a property of the submissions but rather a property of the "send mode" Collect is currently in, though. Adding another field to `SmsSubmission` means more state to keep in sync and generally more complexity.

How about in `bindView`, if Collect is currently set to send by SMS, then do the work to show SMS status, otherwise just show the status message from the database. That is, change the `displaySmsStatus` condition to "send via SMS preference set and model != null". Is there a reason you didn't go that direction?

Also please combine the two blocks that have the same condition and are one after the other (https://github.com/jd-alexander/collect/blob/8a0877d0c8b0f7035db03ea7d609f4648b196b8b/collect_app/src/main/java/org/odk/collect/android/adapters/InstanceUploaderAdapter.java#L113 and https://github.com/jd-alexander/collect/blob/8a0877d0c8b0f7035db03ea7d609f4648b196b8b/collect_app/src/main/java/org/odk/collect/android/adapters/InstanceUploaderAdapter.java#L123)

jd-alexander commented:
Yes, I didn't use that approach because the last status wouldn't be shown when the transport is switched. eg. If I was sending via SMS and it failed and I switched to the internet transport it would still show the status from the database which would be SMS related, it just wouldn't show a proper progress since that data would be tied to a `SmsSubmission` and other items such as the stop button rely on knowing if a SMS submission was just done.

lognaturel commented:
> If I was sending via SMS and it failed and I switched to the internet transport it would still show the status from the database which would be SMS related, it just wouldn't show a proper progress since

Is that critical information?

> stop button 

Do you mean to cancel a submission? This would be relevant in the case that a submission is sending via SMS while the user changes the transport, right? Any other cases?

jd-alexander commented:
> IfI was sending via SMS and it failed and I switched to the internet transport it would still show the status from the database which would be SMS related, it just wouldn't show a proper progress since
Is that critical information?

Honestly, no it isn't I can simply just utilize the concept Yaw spoke of about using 50% for failure/in progress and 100% for success.

> stop button
Do you mean to cancel a submission? This would be relevant in the case that a submission is sending via SMS while the user changes the transport, right? Any other cases?

Yes, but I suppose once the user switches to another transport that shouldn't really matter. 

I thought adding the flag wasn't alot of complexity but now I am seeing your point of view clearly. I will make the relevant changes then.

lognaturel commented:
I hope I'm not sending you down a bad path! I'm trying to think through all the different state changes and it seems sufficient and simpler. There will be more weirdness once we allow SMS sending to be configured in the form and therefore can have a mix of submissions via SMS and via HTTP at the same time but I think we can leave that to later. I think the same approach will work -- we'll check if (the form is set to send via SMS or the Collect-wide mode is SMS) and the model is not null.

jd-alexander commented:
> I hope I'm not sending you down a bad path! I'm trying to think through all the different state changes and it seems sufficient and simpler.

Nah, it's all good. It is simpler and if the other cases I described aren't as critical then doing it simpler is better.

> There will be more weirdness once we allow SMS sending to be configured in the form and therefore can have a mix of submissions via SMS and via HTTP at the same time but I think we can leave that to later. I think the same approach will work -- we'll check if (the form is set to send via SMS or the Collect-wide mode is SMS) and the model is not null.

Okay!

jd-alexander committed:
InstanceUploaderAdapter now uses the transport medium of the submission to show it's status.

A submission that was done via SMS will only show that specific status when the transport is set to sms otherwise it will show a generalized status similar to Aggregate/Google implementations. Once no SmsSubmission model is found then the latest status will be shown which would be an internet upload.

jd-alexander committed:
SmsService is now persisting the sent SMS status message to the instances db

Previously an instance was just marked as submitted so the default Aggregate/Google sent message was being persisted. Now the SMS variant is being generated and persisted.

mmarciniak90 commented:
@lognaturel @jd-alexander  I'm not sure that I see expected result in this case:
1. User sets `Send submission via SMS`
2. User sends submission via SMS (status is `Sent as SMS`)
3. User sets `Send submission via WiFi`
4. User opens 'Send Finalized Forms' and selects `Show Sent and Unsent Forms` (status is `Sent as SMS`)
5. User sends a submission to Aggregate (status is `Send on [date]`)
6. User sets `Send submission via SMS`
7. User opens `Send Finalized Forms` and selects `Show Sent and Unsent Forms` 
8. Status is `Sent as SMS` although submission was sent via WiFi as a last

jd-alexander commented:
Yes, that's the expected behavior based on the discussion I had with @lognaturel about this implementation. It loses the ability to show the last status because once the transport is set to `SMS` then that status is shown.

lognaturel commented:
@jd-alexander I thought that if the SMS submission was successful then the model would be discarded and so the `model != null` test would fail. Is that not the case? Is it still in shared preferences? I believe all of shared preferences are read into memory so we have to be really careful to clean that up so users don't run out of heap.

jd-alexander commented:
@lognaturel Yes, that's what was happening until I was running into some errors when the SmsService was deleting the model after submission and then the `NotificationReceiver` was trying to access it to grab data to display the notification. I can circumvent that though, by just sending all the data needed by the receiver via the intent and perform the deletion. Once that's done then the implementation should work fine.

lognaturel commented:
Ok, that will be great to fix! The model should definitely be deleted after submission.

jd-alexander commented:
Sure! Will do.

jd-alexander committed:
SmsSubmission model is now deleted once the submission is complete.

Previously the model was preserved due to it's usage in the NotificationReceiver and the deletion point was within the SmsService. Now with this restructuring the NotificationReceiver is deleting the model after it's complete with it and if the app is in the foreground then the InstanceUploaderList does the deletion.

mmarciniak90 commented:
**Tested with success**

Verified on Android: 6.0, 5.1, 4.4, 4.2, 4.1

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"
